### DISCUSSION TITLE
Partial support for maven SNAPSHOTs

### DISCUSSION THREAD

**Discussion by gsmachado** [community/community#24916] [Packages]
<p>Hello,</p>
<p><a href="https://github.community/t/github-package-snapshot-build-number-not-updating/14689">This answer</a> says that it’s possible to upload/push snapshot packages with the same version (e.g., 0.1-SNAPSHOT).</p>
<p>However, it also says that there was a bug that prevents to push artifacts with -javadoc and -sources extension, somehow. Is that still an issue?</p>
<p>Because I’m trying to upload/push version 3.0.1-SNAPSHOT, and I get the following error:</p>
<pre><code class="lang-auto">* What went wrong:
Execution failed for task ':compiler:publishMavenJavaPublicationToMavenRepository'.
&gt; Failed to publish publication 'mavenJava' to repository 'maven'
   &gt; Could not GET 'https://maven.pkg.github.com/neow3j/neow3j/io/neow3j/compiler/3.0.1-SNAPSHOT/maven-metadata.xml'. Received status code 400 from server: Bad Request
</code></pre>
<p>When I go to fetch the <code>maven-metadata.xml</code> file on the URL from above, I get:</p>
<pre><code class="lang-auto">error retrieving metadata for snapshot file: compiler-3.0.1-20200527.103236-1-sources.jar
</code></pre>
<p>What’s the problem?</p>
<p>Thanks!</p>

**Comment by jcansdale** [comment]
<p>Hi <a class="mention" href="/gsmachado">@gsmachado</a>,</p>
<p>Sorry about the delay in getting back to you!</p>
<aside class="quote no-group" data-username="gsmachado" data-post="1" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 gsmachado:</div>
<blockquote>
<p>However, it also says that there was a bug that prevents to push artifacts with -javadoc and -sources extension, somehow. Is that still an issue?</p>
</blockquote>
</aside>
<p>This issue <em>should</em> now be fixed.</p>
<aside class="quote no-group" data-username="gsmachado" data-post="1" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 gsmachado:</div>
<blockquote>
<p>When I go to fetch the <code>maven-metadata.xml</code> file on the URL from above, I get:</p>
<pre><code class="lang-auto">error retrieving metadata for snapshot file: compiler-3.0.1-20200527.103236-1-sou
</code></pre>
</blockquote>
</aside>
<p>I am able to fetch your <code>maven-metadata.xml</code> using the following command:</p>
<pre><code class="lang-auto">$ curl https://maven.pkg.github.com/neow3j/neow3j/io/neow3j/compiler/3.0.1-SNAPSHOT/maven-metadata.xml -u token:${READ_PACKAGES_TOKEN} | xmllint --format -
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1661  100  1661    0     0   3831      0 --:--:-- --:--:-- --:--:--  3827
&lt;?xml version="1.0"?&gt;
&lt;metadata modelVersion=""&gt;
  &lt;groupId&gt;io.neow3j&lt;/groupId&gt;
  &lt;artifactId&gt;compiler&lt;/artifactId&gt;
  &lt;version&gt;3.0.1-SNAPSHOT&lt;/version&gt;
  &lt;versioning&gt;
    &lt;snapshot&gt;
      &lt;timestamp&gt;20200527.103236&lt;/timestamp&gt;
      &lt;buildNumber&gt;1&lt;/buildNumber&gt;
    &lt;/snapshot&gt;
    &lt;lastUpdated&gt;20200527103248&lt;/lastUpdated&gt;
    &lt;snapshotVersions&gt;
      &lt;snapshotVersion&gt;
        &lt;extension&gt;jar&lt;/extension&gt;
        &lt;value&gt;3.0.1-20200527.103236-1&lt;/value&gt;
        &lt;updated&gt;20200527103237&lt;/updated&gt;
      &lt;/snapshotVersion&gt;
      &lt;snapshotVersion&gt;
        &lt;extension&gt;pom&lt;/extension&gt;
        &lt;value&gt;3.0.1-20200527.103236-1&lt;/value&gt;
        &lt;updated&gt;20200527103238&lt;/updated&gt;
      &lt;/snapshotVersion&gt;
</code></pre>
<p><code>READ_PACKAGES_TOKEN</code> is a PAT with the <code>read:packages</code> scope.</p>
<p>Any chance you could point me to a GitHub Actions workflow that shows this failing?</p>

**Comment by gsmachado** [comment]
<p>Hello <a class="mention" href="/jcansdale">@jcansdale</a></p>
<p>Thanks for your reply, even if took a while. No worries! :wink:</p>
<p>So… I can confirm that now I can successfully publish <code>3.0.1-SNAPSHOT</code> – i.e., update/re-upload artifacts using the same version (like an overwrite). That’s fine!</p>
<p>However, I still don’t get it: even if the repo is public (open-source project), users mandatorily need to use a token with <code>read:packages</code> scope, somehow?</p>
<p>I don’t know if you’re familiar with Java/Gradle ecosystem, but, if I set-up a project with the following <code>gradle.build</code> content, it fails:</p>
<pre><code class="lang-auto">plugins {
    id 'java'
}

group 'org.example'
version '1.0-SNAPSHOT'

repositories {
    mavenCentral()
    maven {
        url "https://maven.pkg.github.com/neow3j/neow3j/"
    }
}

dependencies {
    compile 'io.neow3j:contract:3.0.1-SNAPSHOT'
    testCompile group: 'junit', name: 'junit', version: '4.12'
}
</code></pre>
<p>The output of a simple <code>gradle compileJava</code> on this repository example gives the following output:</p>
<pre><code class="lang-auto">&gt; Could not resolve all files for configuration ':compileClasspath'.
   &gt; Could not resolve io.neow3j:contract:3.0.1-SNAPSHOT.
     Required by:
         project :
      &gt; Could not resolve io.neow3j:contract:3.0.1-SNAPSHOT.
         &gt; Unable to load Maven meta-data from https://maven.pkg.github.com/neow3j/neow3j/io/neow3j/contract/3.0.1-SNAPSHOT/maven-metadata.xml.
            &gt; Could not get resource 'https://maven.pkg.github.com/neow3j/neow3j/io/neow3j/contract/3.0.1-SNAPSHOT/maven-metadata.xml'.
               &gt; Could not GET 'https://maven.pkg.github.com/neow3j/neow3j/io/neow3j/contract/3.0.1-SNAPSHOT/maven-metadata.xml'. Received status code 401 from server: Unauthorized
</code></pre>
<p>If this is really the case, and I got it right, <strong>this is a real problem for the adoption of Maven-related projects</strong>! It will certainly bring lots of frustration, and people will not widely use it. :frowning_face:</p>
<p>People (users of libraries) just want to add the repository URL, add the dependency, and then start using libs/artifacts! That’s it. No generation of tokens. No account on GitHub. No addition configuration. Nothing. It’s open-source software anyway, thus, packages and its metadata (maven-metadata.xml) should be widely accessible. :slight_smile:</p>
<p>Let me know if I misinterpreted something… also, maybe there’s a configuration, somewhere, that I missed.</p>
<p>Thanks!</p>

**Comment by jcansdale** [comment]
<p><a class="mention" href="/gsmachado">@gsmachado</a>,</p>
<aside class="quote no-group" data-username="gsmachado" data-post="3" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 gsmachado:</div>
<blockquote>
<p>However, I still don’t get it: even if the repo is public (open-source project), users mandatorily need to use a token with <code>read:packages</code> scope, somehow?</p>
</blockquote>
</aside>
<p>From what I understand, it’s currently intended more for internal development dependencies than for hosting public packages. Packages can be made accessible to anyone with a GitHub account who is motivated to create a PAT (which isn’t exactly public).</p>
<p>It is possible to create a <code>read:packages</code> token on behalf of users that might want to build your project. There are however a couple of complications:</p>
<ol>
<li>GitHub will automatically delete tokens that are pushed to public repositories</li>
<li>The <code>read:packages</code> scope gives access packages associated with private as well as public packages</li>
</ol>
<aside class="quote no-group" data-username="gsmachado" data-post="3" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 gsmachado:</div>
<blockquote>
<p>I don’t know if you’re familiar with Java/Gradle ecosystem, but, if I set-up a project with the following <code>gradle.build</code> content, it fails:</p>
</blockquote>
</aside>
<p>I’m still experimenting with Gradle configurations, but what you might be able to do is something like the following:</p><aside class="onebox githubpullrequest" data-onebox-src="https://github.com/jcansdale-test/gradle-java-consume/pull/1">
  <header class="source">

      <a href="https://github.com/jcansdale-test/gradle-java-consume/pull/1" target="_blank" rel="noopener nofollow ugc">github.com/jcansdale-test/gradle-java-consume</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Pull Request">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 12 16" aria-hidden="true"><path d="M11 11.28V5c-.03-.78-.34-1.47-.94-2.06C9.46 2.35 8.78 2.03 8 2H7V0L4 3l3 3V4h1c.27.02.48.11.69.31.21.2.3.42.31.69v6.28A1.993 1.993 0 0 0 10 15a1.993 1.993 0 0 0 1-3.72zm-1 2.92c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zM4 3c0-1.11-.89-2-2-2a1.993 1.993 0 0 0-1 3.72v6.56A1.993 1.993 0 0 0 2 15a1.993 1.993 0 0 0 1-3.72V4.72c.59-.34 1-.98 1-1.72zm-.8 10c0 .66-.55 1.2-1.2 1.2-.65 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/jcansdale-test/gradle-java-consume/pull/1" target="_blank" rel="noopener nofollow ugc">Use a public read:packages token</a>
    </h4>

    <div class="branches">
      <code>jcansdale-test:master</code> ← <code>jcansdale-test:public-token</code>
    </div>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2020-07-20" data-time="08:16:28" data-timezone="UTC">08:16AM - 20 Jul 20 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/jcansdale-test" target="_blank" rel="noopener nofollow ugc">
          <img alt="jcansdale-test" src="https://user-images.githubusercontent.com/11719160/181087729-f2a4ebc3-e72f-469a-a557-5ad08f269881.png" class="onebox-avatar-inline" width="20" height="20">
          jcansdale-test
        </a>
      </div>

      <div class="lines" title="2 commits changed 2 files with 2 additions and 5 deletions">
        <a href="https://github.com/jcansdale-test/gradle-java-consume/pull/1/files" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+2</span>
          <span class="removed">-5</span>
        </a>
      </div>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">### What this PR does

1. Create a PAT with the `read:packages` scope from a s<span class="show-more-container"><a href="https://github.com/jcansdale-test/gradle-java-consume/pull/1" target="_blank" rel="noopener nofollow ugc" class="show-more">…</a></span><span class="excerpt hidden">afe account (e.g. a [machine user](https://developer.github.com/v3/guides/managing-deploy-keys/#machine-users) account)
2. Use `docker run jcansdale/gpr encode &lt;PAT&gt;` to encode the token
3. Add a `credentials` element to the maven details with `username` and `password`
4. `username` can be anything and `password` the encoded string from example `.npmrc` file

For example:

```gradle
repositories {
    maven {
        name = "remote"
        // Adapt the URL for your remote repository
        url = uri("https://maven.pkg.github.com/jcansdale-test/gradle-java-publish")
        credentials { 
            // Use this if the repo requires auth 
            // see https://docs.gradle.org/6.4/userguide/declaring_repositories.html#sec:supported_transport_protocols
            username = "token"
            password = "\u003c\u0050\u0041\u0054\u003e...................................................."
        }
    }
}
```</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<ol>
<li>Create a PAT with the <code>read:packages</code> scope from a safe account (e.g. a <a href="https://developer.github.com/v3/guides/managing-deploy-keys/#machine-users" rel="noopener nofollow ugc">machine user</a> account)</li>
<li>Use <code>docker run jcansdale/gpr encode &lt;PAT&gt;</code> to encode the token (see <a href="https://github.com/jcansdale/gpr" rel="noopener nofollow ugc">here</a>)</li>
<li>Add a <code>credentials</code> element to the maven details with <code>username</code> and <code>password</code>
</li>
<li>
<code>username</code> can be anything and <code>password</code> the encoded string from example <code>.npmrc</code> file</li>
</ol>
<p>For example:</p>
<pre><code class="lang-auto">repositories {
    maven {
        name = "remote"
        // Adapt the URL for your remote repository
        url = uri("https://maven.pkg.github.com/jcansdale-test/gradle-java-publish")
        credentials { 
            // Use this if the repo requires auth 
            // see https://docs.gradle.org/6.4/userguide/declaring_repositories.html#sec:supported_transport_protocols
            username = "token"
            password = "\u003c\u0050\u0041\u0054\u003e...................................................."
        }
    }
}
</code></pre>
<p>I’d be interested to know if something like this would work for you.</p>

**Comment by gsmachado** [comment]
<p>Hey <a class="mention" href="/jcansdale">@jcansdale</a></p>
<p>Thanks for the fast reply! Much appreciated! :slight_smile:</p>
<aside class="quote group-github-employees" data-username="jcansdale" data-post="4" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>From what I understand, it’s currently intended more for internal development dependencies than for hosting public packages. Packages can be made accessible to anyone with a GitHub account who is motivated to create a PAT (which isn’t exactly public).</p>
</blockquote>
</aside>
<p>Oh, ok. This is exactly what I meant.<br>
So, it means that GitHub packages feature is not intended for hosting (real) public packages. Gotcha.</p>
<p>However, this is a use case that is interesting for tons of open source projects out there, including <a href="https://github.com/neow3j/neow3j" rel="nofollow noopener">neow3j</a>. The advantages of publishing to GitHub Packages instead of JFrog/SonaType/etc are obvious: everything in a single platform, GitHub. It’s faster, and there’s a bunch of opportunities to explore there (including metrics of download, etc) in the entire software delivery lifecycle: development, release, distribution, measure.</p>
<p>At the moment, what you propose is not really suitable for widely public packages – because of the necessity of creating a PAT.</p>
<p>Where can I submit a Feature Request for this?</p>

**Comment by jcansdale** [comment]
<aside class="quote no-group" data-username="gsmachado" data-post="5" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 gsmachado:</div>
<blockquote>
<p>At the moment, what you propose is not really suitable for widely public packages – because of the necessity of creating a PAT.</p>
</blockquote>
</aside>
<p>What if you create a <code>read:packages</code> PAT that is distributed with the project? Is the solution proposed in the PR above not an option?</p>
<aside class="quote no-group" data-username="gsmachado" data-post="5" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 gsmachado:</div>
<blockquote>
<p>Where can I submit a Feature Request for this?</p>
</blockquote>
</aside>
<p>This is a known issue and being worked on for the next version of GitHub Packages. I’m afraid I can’t give you a timeline though.</p>

**Comment by gsmachado** [comment]
<aside class="quote group-github-employees" data-username="jcansdale" data-post="6" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>What if you create a <code>read:packages</code> PAT that is distributed with the project? Is the solution proposed in the PR above not an option?</p>
</blockquote>
</aside>
<p>Unfortunately, that’s not an option… because I would need to make that <code>read:packages</code> PAT public on the <code>README.md</code> of our project – developers/users should want to copy and paste the details to configure their <code>gradle.build</code> to import our library and start using it. So, since you mentioned that GitHub removes/deletes tokens that are publicly distributed, then, I conclude that’s not a viable option.</p>
<aside class="quote group-github-employees" data-username="jcansdale" data-post="6" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>This is a known issue and being worked on for the next version of GitHub Packages. I’m afraid I can’t give you a timeline though.</p>
</blockquote>
</aside>
<p>Great! We really want this feature. Let us know once it comes out.</p>
<p>If there’s any link or GitHub issue that I can subscribe to follow the progress of such a feature, let me know!</p>
<p>Thanks a lot for your help and directions. :smile:</p>

**Comment by jcansdale** [comment]
<aside class="quote no-group" data-username="gsmachado" data-post="7" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 gsmachado:</div>
<blockquote>
<p>If there’s any link or GitHub issue that I can subscribe to follow the progress of such a feature, let me know!</p>
</blockquote>
</aside>
<p>There isn’t I’m afraid. It would be nice it there was a public repository where we could track this kind of thing. What do you think <a class="mention" href="/whitneyimura">@whitneyimura</a>?</p>

**Comment by gsmachado** [comment]
<p>Hey <a class="mention" href="/jcansdale">@jcansdale</a> and <a class="mention" href="/whitneyimura">@whitneyimura</a><br>
Any news on this matter (unauthenticated packages)?</p>

**Comment by jcansdale** [comment]
<aside class="quote no-group" data-username="gsmachado" data-post="9" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 gsmachado:</div>
<blockquote>
<p>Any news on this matter (unauthenticated packages)?</p>
</blockquote>
</aside>
<p>This was discussed just yesterday. I’m cautiously optimistic. :thinking:</p>

**Comment by gsmachado** [comment]
<p>Hey <a class="mention" href="/jcansdale">@jcansdale</a><br>
Hope you’re doing fine! :slight_smile:</p>
<p>Any more optimistic news about downloading unauthenticated packages? <a class="mention" href="/devhawk">@devhawk</a> would be also interested in this.</p>
<p>Maybe, should we create another post specific to this? Or can you point us to where the discussion is happening?</p>

**Comment by jcansdale** [comment]
<p>Hi <a class="mention" href="/gsmachado">@gsmachado</a> :wave:</p>
<aside class="quote no-group" data-username="gsmachado" data-post="11" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 gsmachado:</div>
<blockquote>
<p>Any more optimistic news about downloading unauthenticated packages? <a class="mention" href="/devhawk">@devhawk</a> would be also interested in this.</p>
</blockquote>
</aside>
<p>No official news I’m afraid. Say hi to <a class="mention" href="/devhawk">@devhawk</a> from me. :grinning_face_with_smiling_eyes:</p>
<p>I’ve just created an example Maven project that depends on a public package. You can simply clone the following repository:</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/jcansdale-test/maven-install-public" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:420/420;"><img src="https://avatars.githubusercontent.com/u/60072076?s=400&amp;amp;v=4" class="thumbnail" width="420" height="420"></div>

<h3><a href="https://github.com/jcansdale-test/maven-install-public" target="_blank" rel="noopener nofollow ugc">jcansdale-test/maven-install-public</a></h3>


  <p><span class="label1">Install public Maven package. Contribute to jcansdale-test/maven-install-public development by creating an account on GitHub.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Then install and execute the project like this:</p>
<pre><code class="lang-auto"># install public packages
mvn install

# execute example app
mvn exec:java -D exec.mainClass=com.octokat.app.InstallApp
</code></pre>
<p>If all goes to plan, it should output:</p>
<pre><code class="lang-auto">Hello from github-packages-examples/maven-publish, commit ed957aa2f450708a0ba42ac901f90bac48ebf012, run number 34!
</code></pre>
<p>The trick is adding a <code>repository</code> entry to the <code>pom.xml</code> like <a href="https://github.com/jcansdale-test/maven-install-public/blob/0182c03597e5e34bfd7e8f72065967db920f0034/pom.xml#L16" rel="noopener nofollow ugc">this</a>:</p>
<pre><code class="lang-xml">&lt;repository&gt;
  &lt;id&gt;github-public&lt;/id&gt;
  &lt;url&gt;https://public:&amp;#56;538964239ca66dda0ba60738699571d1688c709@maven.pkg.github.com/github-packages-examples/*&lt;/url&gt;
&lt;/repository&gt;
</code></pre>
<ul>
<li>
<code>github-public</code> is a unique repository ID. I’m deliberately not using <code>github</code> because the PAT would be overwritten by the <code>actions/setup-java@v1</code> action.</li>
<li>
<code>public</code> is a dummy username (it can be anything).</li>
<li>
<code>&amp;#56;538964239ca66dda0ba60738699571d1688c709</code> is a PAT with the <code>read:packages</code> scope. The first letter is XML encoded to evade auto-deletion of scoped PATs. :wink:
</li>
</ul>
<p>Would something like that work for you?</p>
<p>I’m sure you know more about Maven than me, so please let me know how the example could be improved!</p>

**Comment by gsmachado** [comment]
<p>Cool, <a class="mention" href="/jcansdale">@jcansdale</a>! Yes, <a class="mention" href="/devhawk">@devhawk</a> and I are working in the same community/project (everything is open source). :rocket: We already talked about you being involved in this thread the other day. :grinning_face_with_smiling_eyes:</p>
<p>Thanks for the example! I’m actually using Gradle, but it’s the same logic/structure, yes.</p>
<p>What bothers me a bit here is to release the PAT publicly (even if it’s just read-only). So, a couple of questions:</p>
<ul>
<li>This PATH is associated with an account, and maybe it gets abused, e.g., to widely crawl GitHub things. This could lead to an account suspension, right?</li>
<li>Is there a way to generate a PAT from an organization? Then I would feel a bit more comfortable – since I don’t have to generate from my own individual account.</li>
</ul>
<p>Let me know!</p>
<p>I really appreciate your time and discussing things around here. Really helpful! Keep up the great work.</p>

**Comment by jcansdale** [comment, marked as answer]
<aside class="quote no-group" data-username="gsmachado" data-post="13" data-topic="115628">
<div class="title">
<div class="quote-controls"></div>
 gsmachado:</div>
<blockquote>
<p>Thanks for the example! I’m actually using Gradle, but it’s the same logic/structure, yes.</p>
</blockquote>
</aside>
<p>If you’re using Gradle, <a class="mention" href="/0ffz">@0ffz</a> wrote a plugin which might help. See here:</p><aside class="quote quote-modified" data-post="104" data-topic="14407">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/download-from-github-package-registry-without-authentication/14407/104">Download from Github Package Registry without authentication</a> <a class="badge-wrapper  bullet" href="/c/code-to-cloud/github-packages/43"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #005cc5;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Any questions related to GitHub Packages and how to manage your packages; upload, download, and delete. Happy to see questions about our help docs and the core set of clients and services we support but also questions about configuring and using alternate clients are welcome.">GitHub Packages</span></a>
  </div>
  <blockquote>
    I made a gradle plugin: <a href="https://github.com/0ffz/gpr-for-gradle" class="inline-onebox" rel="noopener nofollow ugc">GitHub - 0ffz/gpr-for-gradle: Gradle plugin for adding maven repos on Github Packages in one line</a> to aid with this a while back, though I was trying to fix up some not-so-clean Groovy syntax because I’m unfamiliar with Groovy and chose to write it in Kotlin. 
Either way, it makes adding a package repo a one-liner but you dont get to customize it much if you’re using Groovy. The syntax looks like this: 
repositories {
    maven githubPackage.invoke("owner/repo")
}

Apparen…
  </blockquote>
</aside>

<blockquote>
<ul>
<li>This PATH is associated with an account, and maybe it gets abused, e.g., to widely crawl GitHub things. This could lead to an account suspension, right?</li>
</ul>
</blockquote>
<p>This could lead to the PAT being rate limited. I don’t think it would lead to account suspension.</p>
<blockquote>
<ul>
<li>Is there a way to generate a PAT from an organization? Then I would feel a bit more comfortable – since I don’t have to generate from my own individual account.</li>
</ul>
</blockquote>
<p>Unfortunately, only users can generate PATs. What I’d recommend at the moment is create a <a href="https://docs.github.com/en/developers/overview/managing-deploy-keys#machine-users" rel="noopener nofollow ugc">machine-user</a> account and generate a <code>read:packages</code> PAT from that. The PAT will then only be able to access public packages.</p>

**Comment by yogurtearl** [comment]
related: https://github.com/gradle/gradle/issues/13330
